src/main/java/com/rocketchat/core/JSONDecoderStrategy.java

```diff
@@ -0,0 +1,28 @@
+package com.rocketchat.core;
+
+import com.google.gson.Gson;
+import com.google.gson.JsonSyntaxException;
+
+import java.lang.reflect.Type;
+import java.util.Optional;
+
+final public class JSONDecoderStrategy implements DecoderStrategy {
+
+    private Gson gson;
+
+    public JSONDecoderStrategy(Gson gson){
+        this.gson = gson;
+    }
+
+
+    @Override
+    public <T> Optional<T> fromJson(String value, Class<T> type) {
+        Optional<T> object = Optional.empty();
+        try {
+            object = gson.fromJson(value, (Type) type);
```

Comment:
Esto te va a tirar `CastException`, ese casteo explícito que haces está mal porque estás forzando en tiempo de compilación a que los tipos entre el valor de la derecha y la variable sean iguales, el json no es un Optional en sí, es de tipo T, entonces lo que querés hacer es:

```suggestion
            object = Optional.ofNullable(gson.fromJson(value, type));
```